contracts/smart-contract-wallet/modules/SessionKeyModule.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {BaseAuthorizationModule, UserOperation, ISignatureValidator} from "./BaseAuthorizationModule.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@account-abstraction/contracts/core/Helpers.sol";
+import "./SessionValidationModules/ISessionValidationModule.sol";
+
+struct SessionStorage {
+    bytes32 merkleRoot;
+}
+
+contract SessionKeyManager is BaseAuthorizationModule {
+    /*
+     * @dev mapping of Smart Account to a SessionStorage
+     * Info about session keys is stored as root of the merkle tree built over the session keys
+     */
+    mapping(address => SessionStorage) internal userSessions;
+
+    function getSessionKeys(
+        address smartAccount
+    ) external view returns (SessionStorage memory) {
+        return userSessions[smartAccount];
+    }
+
+    function setMerkleRoot(bytes32 _merkleRoot) external {
+        _setSessionData(msg.sender, _merkleRoot);
+    }
+
+    /**
+     * @dev validates userOperation
+     * @param userOp User Operation to be validated.
+     * @param userOpHash Hash of the User Operation to be validated.
+     * @return sigValidationResult 0 if signature is valid, SIG_VALIDATION_FAILED otherwise.
+     */
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external view virtual returns (uint256) {
+        SessionStorage storage sessionKeyStorage = _getSessionData(msg.sender);
+        (bytes memory moduleSignature, ) = abi.decode(
+            userOp.signature,
+            (bytes, address)
+        );
+        (
+            uint48 validUntil,
+            uint48 validAfter,
+            address sessionValidationModule,
+            bytes memory sessionKeyData,
+            bytes32[] memory merkleProof,
+            bytes memory sessionKeySignature
+        ) = abi.decode(
+                moduleSignature,
+                (uint48, uint48, address, bytes, bytes32[], bytes)
+            );
+
+        bytes32 leaf = keccak256(
+            abi.encodePacked(
+                validUntil,
+                validAfter,
+                sessionValidationModule,
+                sessionKeyData
+            )
+        );
+        if (
+            !MerkleProof.verify(merkleProof, sessionKeyStorage.merkleRoot, leaf)
+        ) {
+            revert("SessionNotApproved");
```

Comment:
can validateUserOp revert?

Reply:
I see it may revert with AA23 reverted but should we just treat it as failed sig?

Reply:
I was thinking about it and concluded that this error doesn't relate to the signature itself, so I decided to keep the revert here as well as keep reverts in ERC20SessionValidationModule.
https://github.com/bcnmy/scw-contracts/blob/19d046210f70557ccb31f8564658778b63890e42/contracts/smart-contract-wallet/modules/SessionValidationModules/ERC20SessionValidationModule.sol#L34-L56

However, just returning SIG_VALIDATION_FAILED can be an option as well

Reply:
hmm to think about it, it makes sense since it's not technically signature failure. not sure how bundlers treat AA23 and if this could be a blocker for estimating user operations going on new ownerless account

Reply:
If we return SIG_VALIDATION_FAILED from validateUserOp, EntryPoint will revert with `FailedOp(opIndex, "AA24 signature error")`
If we revert in validateUserOp, EntryPoint will revert with `FailedOp(opIndex, string.concat("AA23 reverted: ", revertReason))`

In both cases, this is the revert from EP with the same `error FailedOp(uint256 opIndex, string reason)`
Just with different arguments

so, for the bundlers it should be the same, it is just validation that fails

Reply:
they have this 
```
catch Error(string memory revertReason) {
            revert FailedOp(opIndex, string.concat("AA23 reverted: ", revertReason));
        } catch {
            revert FailedOp(opIndex, "AA23 reverted (or OOG)");
```
in EntryPoint.sol, so they expect  that `validateUserOp` may revert, not only return SIG_VALIDATION_FAILED

Reply:
> If we return SIG_VALIDATION_FAILED from validateUserOp, EntryPoint will revert with `FailedOp(opIndex, "AA24 signature error")` If we revert in validateUserOp, EntryPoint will revert with `FailedOp(opIndex, string.concat("AA23 reverted: ", revertReason))`
> 
> In both cases, this is the revert from EP with the same `error FailedOp(uint256 opIndex, string reason)` Just with different arguments
> 
> so, for the bundlers it should be the same, it is just validation that fails

yes I also checked in entry point code. in case of paymaster it is similar it would anyway revert in case of SIG_VALIDATION_FAILED as well with AA34 signature error and if validatePaymasterUserOp reverts it would revert with AA33 + error. 

the idea is and this is what I was looking for - it should not revert in simulateHandleOp method in the case it is merely SIG_VALIDATION_FAILED because the purpose is, for bundlers to be able to still estimate gas limits with dummy signatures.